### ISSUE TITLE
Handle other spies by default

### ISSUE THREAD

**Issue by posva** [posva/vue-router-mock#143]
- Handle other global spies by default (apart from jest) like sinon which is used in peeky or cypress spies
- Find a way to set the TS type. Is it possible to have this automatic? If not, allowing the user to extend an interface seems like a good middle point

**Comment by cexbrayat** [comment]
I was going to ask a similar question: is it possible to make it work with `vitest` for example?
The error message complains about `createSpy`, but I can't find how to pass it (as we do in Pinia)?

**Comment by posva** [comment]
Not yet, but I do need to add this

**Comment by cexbrayat** [comment]
I hacked a working version to unlock our migration to vitest, and ended up with something like this (to mimick what you're currently doing):

```
const router = createRouterMock({
  createSpy: fn => {
    const spy = vi.fn(fn);
    return [spy, () => spy.mockClear()];
  }
});
```

Do you imagine something like this, or maybe (haven't tried):

```
const router = createRouterMock({
  createSpy: vi.fn(),
  clearSpy: spy => spy.mockClear()
});
```
?

I can work on a PR to add this, if you give me some guidance.

Another path would be to automatically handle vitest, which would be nice, but I'm not sure this is as easily doable as for sinon or jest...

**Comment by posva** [comment]
maybe grouping spy options in an object:

```js
spy: {
  create: ...
  restore: ...
}
```

To me the most interesting part is allowing a way to handle the types, probably by extending a global interface as I don't think there is a way to check for global types in TS? (`typeof vi extends Record<any, any> ? vi.Spy : handle others` doesn't work, does it?). IMO we should automatically handle jest, sinon (as a global), cypress, and vitest

Feel free to work on this, I will come back to this in a week or two and will be able to bring any guidance too!

**Comment by cexbrayat** [comment]
I'll put together a first version that unblocks the use use of other test fwk, with a `spy` option 👍

For the type checking part, I'm not sure, we'll need to try, but that might not be easy...

It would be awesome to handle vitest out of the box, but unlike the others, it has to be imported (no globals AFAIK).
So we probably need to use a dynamic import and a try/catch to see if vitest is being used in the project? But that makes `createSpy` `async`... Or do you have a better idea?

**Comment by posva** [comment]
We definitely don't want to make `spy.create` async. In that case we might not be able to support vitest out of the box. But I think it's fine, since the setup code can be put in one single place (like setupTestFiles or similar in jest)

**Comment by cexbrayat** [comment]
@posva I took a stab at it, and opened #144 : it is a first step allowing to pass a `spy` option. Even if not super pretty, this will unblock people that want to use other test frameworks

**Comment by posva** [comment]
<img width="656" alt="Screenshot 2022-01-21 at 18 22 32" src="https://user-images.githubusercontent.com/664177/150571807-22d53178-ccc7-435d-ac85-4fc91f2edd80.png">

Seems to be good, even with typing!

**Comment by cexbrayat** [comment]
Nice! 👌

**Comment by posva** [comment]
Published in 0.1.4. Let me know if it works. For adding automatic types for vitest I wrote this:

```ts
import { createRouterMock } from 'vue-router-mock'
import { JestMockCompatFn, vi } from 'vitest'

declare module 'vue-router-mock' {
  export interface RouterMockSpy<Fn> {
    spy: JestMockCompatFn<Parameters<Fn>, ReturnType<Fn>>
  }
}
```
